cohttp-lwt/src/connection_cache.ml

```diff
@@ -0,0 +1,150 @@
+        Connection.close connection;
+        Lwt.return_unit
+    end;
+    res
+end
+
+module Make (Connection : S.Connection) (Sleep : S.Sleep with type 'a promise = 'a Lwt.t) : sig
+  include S.Connection_cache with module IO = Connection.Net.IO
+
+  val create :
+    ?ctx:Connection.Net.ctx ->
+    ?keep:int64 ->
+    ?retry:int ->
+    ?parallel:int ->
+    ?depth:int ->
+    unit -> t
+end
+=
+struct
+  module Net = Connection.Net
+  module IO = Net.IO
+  open IO
+
+  type ctx = Net.ctx
+
+  exception Retry = S.Retry
+
+  module Endp_map = Map.Make (Net.Endp)
+
+  (* type t = < request : ?body:Body.t -> Cohttp.Request.t -> (Cohttp.Response.t * Body.t) IO.t > *)
+  type t =
+    { mutable cache : Connection.t list Endp_map.t
+    ; ctx : ctx
+    ; keep : int64
+    ; retry : int
+    ; parallel : int
+    ; depth : int
+    }
+
+  (*
+   * Select from existing connections by hostname, not IP.
+   * One hostname may resolve to multiple IP addresses.
+   * In such a case we want only one connection.
+   * The downside is that some resolver redirections may break.
+   *)
+
+  let create ?(ctx = Net.default_ctx) ?(keep=60_000_000_000L) ?(retry=2) ?(parallel=4) ?(depth=100) () =
+    { cache = Endp_map.empty
+    ; ctx ; keep ; retry ; parallel ; depth
+    }
+
+  let rec get_connection self endp =
+    let finalise connection =
+      let conns = Endp_map.find endp self.cache in
+      let conns = List.filter ((!=) connection) conns in
+      if conns = []
+      then self.cache <- self.cache |> Endp_map.remove endp
+      else self.cache <- self.cache |> Endp_map.add endp conns;
+      Lwt.return_unit
+    in
+    let create () =
+      let connection = Connection.create ~finalise ~ctx:self.ctx endp
+      and timeout = ref Lwt.return_unit in
+      let rec busy () =
+        Lwt.cancel !timeout;
+        if Connection.length connection = 0 then begin
+          timeout :=
+            Sleep.sleep_ns self.keep >>= fun () ->
+            Connection.close connection; (* failure is ignored *)
+            Lwt.return_unit
+        end;
+        Lwt.on_termination (Connection.notify connection) busy
+      in busy ();
+      connection
+    in
+    match Endp_map.find_opt endp self.cache with
+    | None ->
+      let connection = create () in
+      self.cache <- self.cache |> Endp_map.add endp [connection];
+      Lwt.return connection
+    | Some conns ->
+      let rec search length = function
+        | a :: [] -> a, length + 1
+        | a :: b :: tl
+          when Connection.length a < Connection.length b ->
+          search (length + 1) (a :: tl)
+        | _ :: tl ->
+          search (length + 1) tl
+        | [] -> assert false
+      in
+      match search 0 conns with
+      | shallowest, _ when Connection.length shallowest = 0 ->
+        Lwt.return shallowest
+      | _, length when length < self.parallel ->
+        let connection = create () in
+        self.cache <- self.cache |> Endp_map.add endp (connection :: conns);
```

Comment:
```suggestion
        self.cache <- Endp_map.add endp (connection :: conns) self.cache;
```
Why the indirection?

Reply:
I wanted it to read similar `self.cache += endp`. I'm not feeling very attached to it, so I can change it if its stupid.

Reply:
I don't have such a strong opinion, but I think in ocaml < 4.13 this is slower than function application, and I think it reads more easily the non piped way